src/main/java/io/jenkins/plugins/sshbuildagents/ssh/mina/ConnectionImpl.java

```diff
@@ -0,0 +1,363 @@
+            permissions.add(PosixFilePermission.OWNER_WRITE);
+            scp.upload(bytes, remotePath, permissions, null);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void setServerHostKeyAlgorithms(String[] algorithms) {}
+
+    /** {@inheritDoc} */
+    @Override
+    public void setTCPNoDelay(boolean tcpNoDelay) {
+        this.tcpNoDelay = tcpNoDelay;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public ClientSession connect() throws IOException {
+        initClient();
+        if (isSession() == false) {
+            for (int i = 0; i <= maxNumRetries; i++) {
+                try {
+                    return connectAndAuthenticate();
+                } catch (Exception ex) {
+                    String message = getExMessage(ex);
+                    if (maxNumRetries - i > 0) {
+                        println(
+                                stderr,
+                                "SSH Connection failed with IOException: \""
+                                        + message
+                                        + "\", retrying in "
+                                        + retryWaitTime
+                                        + " seconds."
+                                        + " There are "
+                                        + (maxNumRetries - i)
+                                        + " more retries left.");
+                    }
+                }
+                waitToRetry();
+            }
+            throw new IOException("Max number or reties reached.");
+        }
+        return session;
+    }
+
+    /**
+     * @return True is the session is authenticated and open.
+     */
+    private boolean isSession() {
+        return session != null && session.isAuthenticated() && session.isOpen();
+    }
+
+    /**
+     * Connets to the SSH service configured and authenticate
+     *
+     * @return Returns a ClientSession connected and authenticated.
+     * @throws IOException in case of error.
+     * @throws InterruptedException
+     */
+    private ClientSession connectAndAuthenticate() throws IOException, InterruptedException {
+        ConnectFuture connectionFuture = client.connect(this.credentials.getUsername(), this.host, this.port);
+        connectionFuture.verify(this.timeoutMillis);
+        session = connectionFuture.getSession();
+        var authenticator = SSHAuthenticator.newInstance(session, credentials);
+        authenticator.authenticate(TaskListener.NULL);
+        return session;
+    }
+
+    /** Initialize the SSH client. It reuses the client if it exists. */
+    private void initClient() {
+        if (client == null) {
+            client = SshClient.setUpDefaultClient();
+            CoreModuleProperties.WINDOW_SIZE.set(client, WINDOW_SIZE);
+            CoreModuleProperties.TCP_NODELAY.set(client, tcpNoDelay);
+            CoreModuleProperties.HEARTBEAT_REQUEST.set(client, "keepalive@jenkins.io");
```

Comment:
using KeepAliveHandler would have the advantage about testing the application (ssh) level and not only low level 
see commit here https://github.com/olamy/ssh-agents-plugin/commit/f3f3b6d3d32ce46226cb1aa169c73737e6b92390 
the KeepAlive have some benefits such:
- SSH protocol-level compliant so the server is expected to respond.
- will help detect unreachable servers 
- MINA can close the session automatically if the server doesn’t respond.

Reply:
You can configure the keepalive option in the SSH server. I need to verify whether enabling it in the client affects the configuration you set on the server.

Reply:
> especially this
> 
> ```
> This configuration only ensures that the server does not terminate the session due to no traffic. If the incoming traffic from the server may also suffer from long "quiet" periods, one runs the risk of a client time-out.
> ```

Which is not entirely true. It also helps to **detect** a broken (for whatever reason) connection, and **THAT** is the important part. Regardless, a server can terminate the the connection at any time on the application level.

Reply:
> > especially this
> > ```
> > This configuration only ensures that the server does not terminate the session due to no traffic. If the incoming traffic from the server may also suffer from long "quiet" periods, one runs the risk of a client time-out.
> > ```
> 
> Which is not entirely true. It also helps to **detect** a broken (for whatever reason) connection, and **THAT** is the important part. Regardless, a server can terminate the the connection at any time on the application level.

not sure to understand? Do you mean documentation from Apache Mina is wrong?
My point is we can have both as it doesn't hurt much (or having it optional but activable via SystemProperties).